testing/test-setup/src/lib/extend/ui-logger.matcher.ts

```diff
@@ -0,0 +1,180 @@
+import { cliui } from '@poppinss/cliui';
+import type { SyncExpectationResult } from '@vitest/expect';
+import { expect } from 'vitest';
+import {
+  type LogLevel,
+  extractLevel,
+  extractMessage,
+  hasExpectedMessage,
+  messageContains,
+} from './ui-logger.matcher.utils';
+
+type CliUi = ReturnType<typeof cliui>;
+
+export type CustomUiLoggerMatchers = {
+  toHaveLoggedMessage: (expected: string) => void;
+  toHaveLoggedNthMessage: (nth: number, expected: string) => void;
+  toHaveLoggedLevel: (expected: LogLevel) => void;
+  toHaveLoggedNthLevel: (nth: number, expected: LogLevel) => void;
+  toHaveLoggedMessageContaining: (expected: string) => void;
+  toHaveLoggedNthMessageContaining: (nth: number, expected: string) => void;
+  toHaveLogged: () => void;
+  toHaveLoggedTimes: (times: number) => void;
+};
+
+expect.extend({
+  toHaveLoggedMessage: assertMessageLogged,
+  toHaveLoggedNthMessage: assertNthMessageLogged,
+  toHaveLoggedLevel: assertLevelLogged,
+  toHaveLoggedNthLevel: assertNthLevelLogged,
+  toHaveLoggedMessageContaining: assertMessageContaining,
+  toHaveLoggedNthMessageContaining: assertNthMessageContaining,
+  toHaveLogged: assertLogs,
+  toHaveLoggedTimes: assertLogCount,
+});
+
+function assertMessageLogged(
+  actual: CliUi,
+  expected: string,
+): SyncExpectationResult {
+  const messages = actual.logger
+    .getRenderer()
+    .getLogs()
+    .map(({ message }) => extractMessage(message));
+
+  const pass = messages.some(msg => hasExpectedMessage(expected, msg));
+  return {
+    pass,
+    message: () =>
+      pass
+        ? `Expected not to have logged: ${expected}`
+        : `Expected to have logged: ${expected}}`,
+  };
+}
+
+function assertNthMessageLogged(
+  actual: CliUi,
+  nth: number,
+  expected: string,
+): SyncExpectationResult {
+  const messages = actual.logger
+    .getRenderer()
+    .getLogs()
+    .map(({ message }) => extractMessage(message));
+
+  const pass = hasExpectedMessage(expected, messages[nth - 1]);
+  return {
+    pass,
+    message: () =>
+      pass
+        ? `Expected not to have logged at position ${nth}: ${expected}`
+        : `Expected to have logged at position ${nth}: ${expected}`,
+  };
+}
+
+function assertLevelLogged(
+  actual: CliUi,
+  expected: LogLevel,
+): SyncExpectationResult {
+  const levels = actual.logger
+    .getRenderer()
+    .getLogs()
+    .map(({ message }) => extractLevel(message));
```

Comment:
The `actual.logger.getRenderer().getLogs().map(({ message }) => message)` logic could be extracted to a helper function.